Commpath

rountree opened the pull request:
Renames `location` environment variables, fields, and enums to `commpath`.

Builds with --with-rm=[flux|serial|slurm].  Can't build slurm-plugin, but that's not due to these changes.

rountree-alt committed:
Cachepath:  parse_loc utilities.

rountree-alt committed:
Cachepath:  remove/rename [orig_]location.

rountree-alt committed:
Cachepath:  Configure-time support.

rountree-alt committed:
Cachepath:  Internal messaging for path resolution

rountree-alt committed:
Cachepath:  Adds cobo_allreduce()

rountree-alt committed:
Cachepath:  Adds parameters to config_mgr

rountree-alt committed:
Cachepath:  adds flux parameter support

rountree-alt committed:
Cachepath:  Adds logging support.

rountree-alt committed:
Cachepath:  Removes out-of-root cleanup checks.

rountree-alt committed:
Cachepath:  Set of small, miscellaneous patches.

rountree-alt committed:
Fixes per Matt's comments.

rountree-alt committed:
Single source of truth for client cachepath.

Previously,

chosen_realized_cachepath was copied into
set_intercept_readlink_cachepath()

chosen_realized_cachepath and chosen_parsed_cachepath were copied
into set_should_intercept_cachepath()

This PR removes both setter functions and makes the original
pointers global.

rountree-alt committed:
Comments the cachepath variables.

rountree-alt committed:
Removes internal vars from spindle_launch.h

Removes chosen_cachepath and cachepath_bitindex from
  spindle_launch.h

Updates initialization of matching variables in ldcs_process_data.

determineValidCachePaths() moved from spindle_be.cc to
  ldcs_audit_server_process.c to get ldcs_process_data visibility.

Added #include "parseloc.h" to ldcs_audit_server_process.c to get
  declaration of determineValidCachePaths().

Relocated "parseloc.h" to src/util so ldcs_audit_server_process.c
  could find it.

Trued up signedness of types caused my making "parseloc.h" more
  visible, e.g., cachepath_bitidx is now uint64_t everywhere.

rountree-alt committed:
Client cachepath message now uses single response.

The three-message-reply response is now a single message with
two strings.  The symbolic version of the cachepath is no longer
communicated as it was not being used.

rountree-alt committed:
Removes assert(0) in network error paths.

rountree-alt committed:
Renames ldcs_audit_server_md_consensus().

New name is ldcs_audit_server_md_allreduce_AND().

If we get to the point where we're using other allreduce operations
we can solve the problem of duplicating the op list in md-land and
cobo-land.  For now, we're only using one op in md-land, so the
op can go into the function name.

rountree-alt committed:
Adds explicit enum values to CmdlineShortOptions.

rountree-alt committed:
Return instead of exit on network errors.

rountree-alt committed:
Apply rename to configuration and parameters.

rountree-alt committed:
Renaming location variables/fields to commpath.

rountree-alt committed:
Renames enums.

rountree commented:
Found a failure case.  Hold off on review for now.

rountree-alt committed:
Use strdup() for commpath instead of stack var.

Unlikely it would ever make a difference, but this is much more
correct.

rountree-alt committed:
Adds LDCS_COMMPATH

rountree-alt committed:
Restores checkLinkForLeak() to test_driver.c

rountree-alt committed:
Replacing "location" with "commpath" as needed.

src/client/beboot/spindle_bootstrap.c
    Moved orig_location from static global to local
    Renamed symbolic_location to symbolic_commpath
    Renamed orig_location to orig_commpath
    Renamed location to commpath
    Renamed LDCS_LOCATION to LDCS_COMMPATH

src/client/client/client.c
    Renamed LDCS_LOCATION to LDCS_COMMPATH
    Renamed location to commpath

rountree-alt committed:
Continues location rename.

src/client/client_comlib/client_api.c
    Added setenv("LDCS_CHOSEN_PARSED_CACHEPATH", local_cpc);

testsuite/test_driver.c
    Replaced LDCS_LOCATION and LDCS_ORIG_LOCATION checks for cachepath with
        LDCS_CHOSEN_PARSED_CACHEPATH
    Replaced spindle_loc with cachepath

rountree-alt committed:
Fixes -Wsign-compare warning in new code.

rountree-alt committed:
LDCS_CHOSEN_PARSED_CACHEPATH set in bootstrap.

rountree-alt committed:
Updates test_driver.c to ignore FIFO files.

All tests pass with both distinct and identical
commpaths/cachepaths.

rountree-alt committed:
Sets TMPDIR=/tmp in each Dockerfile

Additionally populates /etc/environment just in case ssh is used
to set up the servers.

nchaimov commented on configure:
```diff
@@ -846,7 +846,8 @@ enable_libtool_lock
 enable_maintainer_mode
 with_default_port
 with_default_num_ports
-with_localstorage
```
The configure script will give a warning for unrecognized arguments, so removing the `--with-localstorage` causes builds that still specify `--with-localstorage` to configure without error, but the option is unused. For example, the script in the CI environment that builds Spindle uses `--with-localstorage=/tmp`, and you can see the warning in the CI output: 

```
#18 0.165 configure: WARNING: unrecognized options: --with-localstorage
```
https://github.com/llnl/Spindle/actions/runs/21076617750/job/60620001124?pr=117#step:4:826

I wonder if it would make sense to either:

1. Give an error rather than a warning if `--with-localstorage` is specified, where the error message says that `--with-cachepaths` and `--with-commpath` should be used instead; or,
2. Still accept `--with-localstorage` and use it to set the cachepaths and commpath value.

We should also change the Spindle configure line used in CI to use the new arguments instead of `--with-localstorage`, in `containers/spindle-serial-ubuntu/scripts/build_spindle.sh`, `containers/spindle-flux-ubuntu/scripts/build_spindle.sh`, and `containers/spindle-slurm-ubuntu/testing/scripts/build_spindle.sh`.

rountree replied:
Per Matt, map this to both cache and compaths, although those take precedence.

rountree replied:
I think the best way forward is removing `--with-localstorage`, forcing a configure error if it's used, and updating the CI configure scripts.  The semantics for `--with-localstorage` are slightly different from `--with-cachepaths` in that the latter can accept multiple colon-separated paths, and communicating to the user which option takes precedence and why appears to be more trouble than it's worth for an option that's not widely used.

The output of `configure --help` now looks like this:
```
  --with-localstorage=DIR (obsolete)
                          Use --with-cachepaths and --with-commpath instead.
```

The configure error message looks like this:
```
$ ./configure --with_localstorage=foo
configure: error: unrecognized option: `--with-localstorage=foo'
...
configure: error: requested obsolete option --with-localstorage.  Use --with-cachepaths and --with-commpath instead.
```

nchaimov commented on testsuite/test_driver.c:
```diff
@@ -1284,34 +1283,39 @@ void check_for_path_leaks()
    for (d = readdir(proc_fds); d != NULL; d = readdir(proc_fds)) {
       if (d->d_name[0] == '.')
          continue;
+      // Ignore Spindle fifo files for now.
+      if ( strncmp( "315", d->d_name, 3 ) == 0 )
+          continue;
+      if ( strncmp( "316", d->d_name, 3 ) == 0 )
+          continue;
```
Where do these numbers (315, 316) come from?

rountree replied:
A little above the safe zone for tcsh closings.

nchaimov commented on src/trivial_change:
```diff
@@ -0,0 +1 @@
+Hello, world!
```
I think this was just added to force the CI to run and should be removed.

rountree-alt committed:
Restores --with-localstorage to generate error.

The option is marked as obsolete in configure --help and will
cause an error in configure if it is specified.

As updates the CI configure scripts to use --with-cachepaths and
--with-commpath instead of --with-localstorage.

rountree-alt committed:
Updates spank plugin to use commpath.

Replaces args->location with args->commpath.

rountree-alt committed:
Updates a configure script to use commpath.

rountree-alt committed:
Additional integration for commpath + spank-plugin.

rountree-alt committed:
Fixes two silly bugs.

TMPDIR left out of a docker script, and --cachepath instead of
--cachepaths.

rountree-alt committed:
Testing non-overlapping cache/commpath directories.

rountree-alt committed:
Testing commpath as subdirectory of cachepath.

rountree-alt committed:
Testing cachepath as a subdirectory of commpath.

rountree-alt committed:
Generate debug log artifacts as part of github ci

nchaimov committed:
Fix SPINDLE_DEBUG + SPINDLE_TEST in SPANK plugin

rountree-alt committed:
Adds slurm-plugin to matrix refactor.

rountree-alt committed:
Fixup

rountree-alt committed:
Finalizes logging on spindleRunBE error paths

Also ups the timeout for slurm-plugin test cases.

rountree-alt committed:
Several small fixes, add head node to slurm*

rountree-alt committed:
Restoring non-canonical paths in runner.

rountree-alt committed:
Tarball filenames now allow multiple artifacts

rountree-alt committed:
env doesn't take an array.

rountree-alt committed:
printf, awkwardly.

rountree-alt committed:
One bugfix, one upgrade

rountree-alt committed:
I bet we can't resolve yaml variables in bash

rountree-alt committed:
Remove glob by tarring entire testsuite dir.

rountree-alt committed:
Removes YAML variables from bash tar commands

rountree-alt committed:
Adds ./ in front of testsuite dir.

If this doesn't work, we'll use the canonical dir next.

rountree-alt committed:
Using full paths, serial only, force-"fail"

rountree-alt committed:
Testing hard-coded filename for one DEBUG case

rountree-alt committed:
Gingerly reintroducing a YAML variable.

rountree-alt committed:
Fix typo

rountree-alt committed:
Gingerly restoring environment variables.

rountree-alt committed:
Bring flux back online with lessons learned.

rountree-alt committed:
Pass unique name to upload action.

rountree-alt committed:
Tweaks.

rountree-alt committed:
Unique name fixes.

rountree-alt committed:
slurm and plugin back online

rountree-alt committed:
Removed debugging code.

rountree-alt committed:
Handle absent log in head node cases.
